src/rio.c

```diff
@@ -496,3 +497,138 @@ size_t rioWriteBulkDouble(rio *r, double d) {
+        int broken = 0;
+        for (j = 0; j < r->io.connset.numconns; j++) {
+            if (r->io.connset.state[j] != 0) {
+                /* Skip FDs already in error. */
+                broken++;
+                continue;
+            }
+
+            /* Make sure to write 'count' bytes to the socket regardless
+             * of short writes. */
+            size_t nwritten = 0;
+            while (nwritten != count) {
+                retval = connWrite(r->io.connset.conns[j], p + nwritten, count - nwritten);
+                if (retval <= 0) {
+                    /* With blocking sockets, which is the sole user of this
+                     * rio target, EWOULDBLOCK is returned only because of
+                     * the SO_SNDTIMEO socket option, so we translate the error
+                     * into one more recognizable by the user. */
+                    if (retval == -1 && errno == EWOULDBLOCK) errno = ETIMEDOUT;
+                    break;
+                }
+                nwritten += retval;
+            }
+
+            if (nwritten != count) {
+                /* Mark this FD as broken. */
+                r->io.connset.state[j] = errno;
+                if (r->io.connset.state[j] == 0) r->io.connset.state[j] = EIO;
+            }
+        }
+        if (broken == r->io.connset.numconns) return 0; /* All the FDs in error. */
+        p += count;
+        len -= count;
+        r->io.connset.pos += count;
+    }
+
+    if (doflush) sdsclear(r->io.connset.buf);
+    return 1;
+}
+
+/* Returns 1 or 0 for success/failure. */
+static size_t rioConnsetRead(rio *r, void *buf, size_t len) {
+    UNUSED(r);
+    UNUSED(buf);
+    UNUSED(len);
+    return 0; /* Error, this target does not support reading. */
+}
+
+/* Returns read/write position in file. */
+static off_t rioConnsetTell(rio *r) {
+    return r->io.connset.pos;
+}
+
+/* Flushes any buffer to target device if applicable. Returns 1 on success
+ * and 0 on failures. */
+static int rioConnsetFlush(rio *r) {
+    /* Our flush is implemented by the write method, that recognizes a
+     * buffer set to NULL with a count of zero as a flush request. */
+    return rioConnsetWrite(r, NULL, 0);
+}
+
+static const rio rioConnsetIO = {
+    rioConnsetRead,
+    rioConnsetWrite,
+    rioConnsetTell,
+    rioConnsetFlush,
+    NULL,       /* update_checksum */
+    0,          /* current checksum */
+    0,          /* flags */
+    0,          /* bytes read or written */
+    0,          /* read/write chunk size */
+    {{NULL, 0}} /* union for io-specific vars */
+};
+
+
+void rioInitWithConnset(rio *r, connection **conns, int numconns) {
+    *r = rioConnsetIO;
+    r->io.connset.conns = zmalloc(sizeof(connection *) * numconns);
+    r->io.connset.state = zmalloc(sizeof(int) * numconns);
+
+    for (int i = 0; i < numconns; i++) {
+        connIncrRefs(conns[i]);
+        r->io.connset.conns[i] = conns[i];
+        r->io.connset.state[i] = 0;
+    }
```

Comment:
Sorry I couldn't find our earlier thread on this issue so opening a new one - I don't think we need to track the life cycle of these connections given the contract of rio.

Reply:
I don't see the original commit taking references on the connections either?

![image](https://github.com/valkey-io/valkey/assets/11568491/d7695148-c736-4c60-be10-42771144d456)